File: credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

int count = 15;
long creditNum;
int creditcard[16];
int calculate[8];
int calculateRest[8];
int count2 = 0;
int n = 0;
int n2 = 0;
int calculateCount = 0;
int calculateRestCount = 0;
bool validCard = true;
bool AMEX = false;
bool MASTERCARD = false;
bool VISA = false;
bool AMEXLENGTH = false;
bool VISALENGTH = false;
bool MCLENGTH = false;

int main(void)
{
    creditNum = get_long("Number: ");

    do
    {
        // printf("COUNT TOP: %i\n", count);
        // printf("COUNT2: %i\n", count2);
        // printf("CreditNUM MOD: %ld\n", creditNum % 10);
        // printf("CreditNUM: %ld\n", creditNum);
        // printf("N VALUE: %i\n", n);
        if (count % 2 > 0)
        {

            calculateRest[n2] = creditNum % 10;
            n2++;
        }

        creditcard[count] = creditNum % 10;

        if (count % 2 == 0)
        {
            n++;
            calculate[n] = creditNum % 10;
        }

        count--;
        creditNum /= 10;
    }
    while (creditNum != 0);

    for (int i = 0; i < 16; ++i)
    {
        if (creditcard[0] == 5 && creditcard[1] <= 5 && creditcard[1] != 0)
        {
            MCLENGTH = true;
        }
        if (i == 1 && creditcard[i] == 3 && creditcard[i + 1] == 4)
        {
            AMEXLENGTH = true;
        }
        if (i == 1 && creditcard[i] == 3 && creditcard[i + 1] == 7)
        {
            AMEXLENGTH = true;
        }
        if (i == 3 && creditcard[i] == 4)
        {
            VISALENGTH = true;
        }
        if (creditcard[0] == 4)
        {
            VISALENGTH = true;
            // printf("Visa lenght true\n");
        }
    }

    for (int i = 8; i >= 0; --i)
    {
        // printf("%i", calculate[i]);
        calculate[i] = calculate[i] * 2;
    }

    // printf("\n");

    for (int i = 8; i >= 0; --i)
    {
        // printf("%i", calculate[i]);
        if (calculate[i] > 9)
        {
            calculateCount += calculate[i] % 10;
            calculateCount += calculate[i] / 10;
        }
        else
        {
            calculateCount += calculate[i];
        }
    }

    // printf("\n");

    // printf("Calulate rest: ");

    for (int i = 0; i < 8; ++i)
    {
        // printf("%i", calculateRest[i]);
        calculateRestCount += calculateRest[i];
    }
    // printf("\n");
    calculateCount += calculateRestCount;

    /// printf("Calculate Count: %i\n", calculateCount);

    // printf("CalculateCount Mod : %i\n", calculateCount % 10);

    if (VISALENGTH == true && calculateCount % 10 != 0)
    {

        printf("INVALID\n");
    }

    if (VISALENGTH == true && calculateCount % 10 == 0)
    {
        printf("VISA\n");
    }
    if (AMEXLENGTH == true && calculateCount % 10 == 0)
    {
        printf("AMEX\n");
    }
    else if (AMEXLENGTH != true && calculateCount % 10 < 0)
    {
        printf("INVALID\n");
    }

    if (MCLENGTH == true)
    {
        printf("MASTERCARD\n");
    }
    else if (MCLENGTH != true && calculateCount % 10 < 0)
    {
        printf("INVALID\n");
    }

    // printf("\n");

    if (VISALENGTH != true && MCLENGTH != true && AMEXLENGTH != true)
    {
        printf("INVALID\n");
    }

File: credit.py
# TODO
import string
import sys


def main():
    AMEX = False
    MC = False
    VISA = False
    creditnum = 0

    creditnum = str(getinput())

    """
    print(creditnum)
    """

    numcount = count_digits(creditnum)

    if numcount < 13:
        print("INVALID")

    digits_list = [int(char) for char in creditnum]

    if digits_list[0] == 5 and digits_list[1] < 6 and digits_list[1] > 0:
        MC = True
    if digits_list[0] == 4:
        VISA = True
    if digits_list[0] == 3 and digits_list[1] == 4 or digits_list[1] == 7:
        AMEX = True

    """
    print(digits_list)
    """

    ValidCardCheck = check_credit_num(creditnum, numcount, digits_list)

    if ValidCardCheck == False:
        print("INVALID")
        sys.exit(1)

    if MC == True and ValidCardCheck == True:
        print("MASTERCARD")
    if VISA == True and ValidCardCheck == True:
        print("VISA")
    if AMEX == True and ValidCardCheck == True:
        print("AMEX")
    else:
        print("INVALID")
        sys.exit(1)


def count_digits(num):
    return len(num)


def check_credit_num(creditnum, numcount, digits_list):
    count = 0
    count2 = 0
    i = 0
    n = 0
    list = []
    list2 = []

    if float((numcount / 2) % 2) == 0:
        while i != numcount / 2:
            list.append(str(digits_list[n] * 2))
            n += 2
            i += 1
        list = "".join(list)
        i = 0
        while i < len(list):
            count += int(list[i])
            i += 1

        i = 0
        n = 1
        while i < numcount / 2:
            list2.append(str(digits_list[n]))
            n += 2
            i += 1
        list2 = "".join(list2)
        i = 0
        while i < len(list2):
            count2 += int(list2[i])
            i += 1

    if float((numcount / 2) % 2) > 0:
        n = 1
        if numcount > 13:
            while i < (numcount / 2) - 1:
                list.append(str(digits_list[n] * 2))
                i += 1
                n += 2

            list = "".join(list)
            i = 0
            while i < len(list):
                count += int(list[i])
                i += 1

        if numcount == 13:
            n = 1
            i = 0
            while i < (numcount / 2) - 1:
                list.append(str(digits_list[n] * 2))
                i += 1
                n += 2
            list = "".join(list)
            i = 0
            while i < len(list):
                count += int(list[i])
                i += 1

            i = 0
            n = 0
            while i < (numcount / 2):
                list2.append(str(digits_list[n]))
                n += 2
                i += 1
            list2 = "".join(list2)
            i = 0
            while i < len(list2):
                count2 += int(list2[i])
                i += 1

        if numcount > 13:
            n = 0
            i = 0
            while i < (numcount / 2):
                list2.append(str(digits_list[n]))
                i += 1
                n += 2

            list2 = "".join(list2)
            i = 0
            while i < len(list2):
                count2 += int(list2[i])
                i += 1

    if (count + count2) % 10 == 0:
        return True
    else:
        return False


def getinput():
    number = input("Number: ")
    try:
        number = int(number)
        if number > 13:
            return number
    except ValueError:
        return getinput()


main()
